apps/site/assets/ts/schedule/ScheduleLoader.tsx

```diff
@@ -0,0 +1,241 @@
+}
+
+export const ScheduleLoader = ({
+  component,
+  schedulePageData,
+  updateURL,
+  selectedDirection
+}: Props): ReactElement<HTMLElement> => {
+  const [query] = useQueryParams({
+    // eslint-disable-next-line @typescript-eslint/camelcase
+    "schedule_finder[direction_id]": StringParam,
+    "schedule_finder[origin]": StringParam
+  });
+
+  const changeDirection = (direction: DirectionId): void => {
+    const currentState = getCurrentState();
+    const { modalOpen } = currentState;
+
+    storeHandler({
+      type: "CHANGE_DIRECTION",
+      newStoreValues: {
+        selectedDirection: direction,
+        selectedOrigin: null
+      }
+    });
+    if (!modalOpen) updateURL(direction, null);
+  };
+
+  const changeOrigin = (origin: SelectedOrigin): void => {
+    storeHandler({
+      type: "CHANGE_ORIGIN",
+      newStoreValues: {
+        selectedOrigin: origin
+      }
+    });
+    const currentState = getCurrentState();
+    const { selectedDirection: currentDirection } = currentState;
+    updateURL(currentDirection, origin);
+
+    // reopen modal depending on choice:
+    storeHandler({
+      type: "OPEN_MODAL",
+      newStoreValues: {
+        modalMode: origin === null ? "origin" : "schedule"
+      }
+    });
+  };
+
+  const closeModal = (): void => {
+    storeHandler({
+      type: "CLOSE_MODAL"
+    });
+    // clear parameters from URL when closing the modal:
+    updateURL("", "");
+  };
+
+  React.useEffect(() => {
+    // set initial state for the store when the component mounts
+    let newDirection = selectedDirection;
+    let newOrigin;
+    let modalMode = "origin";
+    let modalOpen = false;
+
+    if (query["schedule_finder[direction_id]"] !== undefined) {
+      newDirection = query["schedule_finder[direction_id]"] === "0" ? 0 : 1;
+    }
+    if (query["schedule_finder[origin]"] !== undefined) {
+      newOrigin = query["schedule_finder[origin]"];
+    }
+
+    const button = document.getElementById("modal-close")!;
+    if (
+      newDirection !== undefined &&
+      newOrigin !== undefined &&
+      button !== undefined
+    ) {
+      modalMode = "schedule";
+      modalOpen = true;
+    }
+
+    storeHandler({
+      type: "INITIALIZE",
+      newStoreValues: {
+        selectedDirection: newDirection,
+        selectedOrigin: newOrigin,
+        modalMode,
+        modalOpen
+      }
+    });
+    // we disable linting in this next line because we DO want to specify an empty array since we want this piece to run only once
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const {
+    route,
+    stops,
+    services,
+    route_patterns: routePatternsByDirection,
+    schedule_note: scheduleNote,
+    today,
+    line_diagram: lineDiagram,
+    shape_map: shapesById,
+    variant: initialSelectedRoutePatternId
+  } = schedulePageData;
+
+  const currentState = getCurrentState();
+
+  if (!!currentState && Object.keys(currentState).length !== 0) {
```

Comment:
@alicia-mbta can you write `currentState.keys`? I think that works, not sure without trying it.